src/riak_kv_vnode.erl

```diff
@@ -1476,147 +1478,162 @@ prepare_put(State=#state{vnodeid=VId,
                              lww=LWW,
                              coord=Coord,
                              robj=RObj,
-                             starttime=StartTime}) ->
+                             starttime=StartTime,
+                             bprops = BProps}) ->
     %% Can we avoid reading the existing object? If this is not an
     %% index backend, and the bucket is set to last-write-wins, then
     %% no need to incur additional get. Otherwise, we need to read the
     %% old object to know how the indexes have changed.
+    IndexBackend = is_indexed_backend(Mod, Bucket, ModState),
+    IsSearchable = ?IS_SEARCH_ENABLED_FOR_BUCKET(BProps),
+    SkipReadBeforeWrite = LWW andalso (not IndexBackend) andalso (not IsSearchable),
+    case SkipReadBeforeWrite of
+        true ->
+            prepare_blind_put(Coord, RObj, VId, StartTime, PutArgs, State);
+        false ->
+            prepare_read_before_write_put(State, PutArgs, IndexBackend, IsSearchable)
+    end.
+
+is_indexed_backend(Mod, Bucket, ModState) ->
```

Comment:
Would it make sense to change the return type for `is_indexed_backed/3` to `indexed_backend | non_indexed_backend`?
And then change the name to `backend_index_type/3`

It would make the logic read nicer:
```
BackendIndexType = backend_index_type(Mod, Bucket, ModState),
...
SkipReadBeforeWrite = LWW andalso (BackendIndexType == non_indexd_backend) andalso ...
```

This would also simplify the `get_index_specs/5` function below as it would not have to name the first argument because the atom says it all.

Might make sense to do the same for `IsSearchable`.